plugins/inputs/firehose/firehose.go

```diff
@@ -0,0 +1,413 @@
+		Handler:      f,
+		ReadTimeout:  time.Duration(f.ReadTimeout),
+		WriteTimeout: time.Duration(f.WriteTimeout),
+		TLSConfig:    f.tlsConf,
+	}
+}
+
+// Stop cleans up all resources
+func (f *Firehose) Stop() {
+	if f.listener != nil {
+		f.listener.Close()
+	}
+	f.wg.Wait()
+}
+
+func (f *Firehose) Init() error {
+	tlsConf, err := f.ServerConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	var listener net.Listener
+	if tlsConf != nil {
+		listener, err = tls.Listen("tcp", f.ServiceAddress, tlsConf)
+	} else {
+		listener, err = net.Listen("tcp", f.ServiceAddress)
+	}
+	if err != nil {
+		return err
+	}
+	f.tlsConf = tlsConf
+	f.listener = listener
+
+	return nil
+}
+
+func (f *Firehose) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	select {
+	case <-f.close:
+		res.WriteHeader(http.StatusGone)
+		return
+	default:
+	}
+
+	if !choice.Contains(req.URL.Path, f.Paths) {
+		res.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	r := &firehoseRequest{req: req}
+
+	if f.AccessKey != "" {
+		if ok := r.authenticate(f.AccessKey); !ok {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), "Unauthorized"))
+			err := r.sendResponse(res)
+			if err != nil {
+				f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+			}
+			return
+		}
+	}
+
+	if err := r.validate(); err != nil {
+		f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		err := r.sendResponse(res)
+		if err != nil {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		}
+		return
+	}
+
+	decodedBytesData, ok := r.decodeData()
+	if !ok {
+		f.Log.Error(formatLog(r.body.RequestID, "Failed to base64 decode record data"))
+		err := r.sendResponse(res)
+		if err != nil {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		}
+		return
+	}
+
+	metrics := make([]telegraf.Metric, 0)
+	for _, bytes := range decodedBytesData {
+		m, err := f.Parse(bytes)
+		if err != nil {
+			f.Log.Error(formatLog(r.body.RequestID, "Unable to parse data"))
+			r.responseStatusCode = http.StatusBadRequest
+			err := r.sendResponse(res)
+			if err != nil {
+				f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+			}
+			return
+		}
+		metrics = append(metrics, m...)
+	}
+
+	if len(metrics) == 0 {
+		once.Do(func() {
+			f.Log.Info(formatLog(r.body.RequestID, internal.NoMetricsCreatedMsg))
+		})
+	}
+
+	commonAttributesHeaderValue := req.Header.Get(commonAttributesHeader)
+	if len(commonAttributesHeaderValue) != 0 && len(f.ParameterTags) != 0 {
+		parameters := make(map[string]interface{})
+		err := json.Unmarshal([]byte(commonAttributesHeaderValue), &parameters)
+		if err != nil {
+			f.Log.Warn(formatLog(r.body.RequestID, commonAttributesHeader+" header's value is not a valid json"))
+		}
+
+		parameters, ok := parameters["commonAttributes"].(map[string]interface{})
+		if !ok {
+			f.Log.Warn(formatLog(r.body.RequestID, "Invalid value for header "+commonAttributesHeader))
```

Comment:
Well I don't think you should simply continue here rather return here...

Reply:
I thought of parameter tags as something optional. So if were able to extract and find them, we'd add to the metrics otherwise we'll ignore. Thus, I'm just emitting a warning log. Now that you mentioned, it does makes sense to maybe respond with bad request status and return here

Reply:
Oh I see, but I would then at least skip the `for` loop below as this is known to not return anything but just burns some cycles...

Reply:
updated